File: src/models/onboarding.ts
import { RepoInfo, TechStack } from '@/types'
import { ArchitectureAnalysis } from './architecture'
import { getOnboardingTemplate } from './knowledge'
import { detectPackageManager, detectDevCommands } from './technologies'

export function generateOnboardingGuide(
  repo: RepoInfo,
  techStack: TechStack,
  architecture: ArchitectureAnalysis
): string {
  const template = getOnboardingTemplate('default')
  const commands = detectDevCommands(repo.dependencyFiles, repo.languages)
  const pm = detectPackageManager(repo.dependencyFiles)

  const langNames = techStack.languages.slice(0, 3).map(l => l.name).join(', ') || 'Your preferred language'
  const linter = techStack.tools.find(t => t.toLowerCase().includes('eslint') || t.toLowerCase().includes('prettier')) || 'the configured linter'
  const mainFrameworks = techStack.frameworks.slice(0, 2).join(', ') || 'your framework of choice'
  const archNotes = architecture.directoryHighlights.slice(0, 4).join('\n')

  const replacements: Record<string, string> = {
    '{{REPO_URL}}': repo.url,
    '{{REPO_NAME}}': repo.name,
    '{{LANGUAGES_DISPLAY}}': langNames,
    '{{INSTALL_COMMAND}}': commands.install,
    '{{DEV_COMMAND}}': commands.dev,
    '{{BUILD_COMMAND}}': commands.build,
    '{{TEST_COMMAND}}': commands.test,
    '{{ARCHITECTURE}}': architecture.primaryPattern,
    '{{STRUCTURE_HIGHLIGHTS}}': archNotes,
    '{{LINTER}}': linter,
  }

  const sections = template.sections.map(section => {
    let result = section
    for (const [key, value] of Object.entries(replacements)) {
      result = result.replace(new RegExp(key.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'g'), value)
    }
    return result
  })

  const hasDocker = Object.keys(repo.dependencyFiles).some(f =>
    ['Dockerfile', 'docker-compose.yml', 'docker-compose.yaml'].includes(f)
  )
  const hasDatabase = techStack.databases.length > 0

  if (hasDocker) {
    sections.push(
      '## Docker (Optional)\n\nIf you prefer using Docker:\n```bash\ndocker-compose up -d\n```'
    )
  }

  if (hasDatabase) {
    sections.push(
      `## Database Setup\n\nThis project uses ${techStack.databases.join(', ')}. Ensure it is running locally or configure the connection in your environment variables.`
    )
  }

  sections.push(
    '## Troubleshooting\n\nIf you encounter issues:\n1. Ensure all prerequisites are installed\n2. Clear your package manager cache and reinstall dependencies\n3. Check that environment variables are properly configured\n4. Look for existing GitHub Issues or open a new one'
  )

  return sections.join('\n\n')
}

File: backend/models/onboarding.py
from __future__ import annotations
from .knowledge import get_onboarding_template
from .technologies import detect_package_manager, detect_dev_commands


def generate_onboarding_guide(
    repo: dict, tech_stack: dict, architecture: dict,
) -> str:
    template = get_onboarding_template()
    commands = detect_dev_commands(repo.get('dependencyFiles', {}), repo.get('languages', {}))
    pm = detect_package_manager(repo.get('dependencyFiles', {}))

    langs = [l.get('name', '') for l in (tech_stack.get('languages', []) or [])[:3]]
    lang_names = ', '.join(langs) or 'Your preferred language'
    tools = tech_stack.get('tools', []) or []
    linter = next((t for t in tools if 'eslint' in t.lower() or 'prettier' in t.lower()), 'the configured linter')
    frameworks = (tech_stack.get('frameworks', []) or [])[:2]
    main_fw = ', '.join(frameworks) or 'your framework of choice'
    arch_notes = '\n'.join((architecture.get('directoryHighlights', []) or [])[:4])

    replacements = {
        '{{REPO_URL}}': repo.get('url', ''),
        '{{REPO_NAME}}': repo.get('name', ''),
        '{{LANGUAGES_DISPLAY}}': lang_names,
        '{{INSTALL_COMMAND}}': commands.get('install', ''),
        '{{DEV_COMMAND}}': commands.get('dev', ''),
        '{{BUILD_COMMAND}}': commands.get('build', ''),
        '{{TEST_COMMAND}}': commands.get('test', ''),
        '{{ARCHITECTURE}}': architecture.get('primaryPattern', architecture.get('primary', 'Standard')),
        '{{STRUCTURE_HIGHLIGHTS}}': arch_notes,
        '{{LINTER}}': linter,
    }

    sections = list(template[0].get('sections', [])) if template else []
    result_sections: list[str] = []
    for section in sections:
        if isinstance(section, str):
            result = section
            for key, value in replacements.items():
                result = result.replace(key, value)
            result_sections.append(result)

    dep_files = repo.get('dependencyFiles', {})
    has_docker = any(f in dep_files for f in ['Dockerfile', 'docker-compose.yml', 'docker-compose.yaml'])
    has_database = len(tech_stack.get('databases', []) or []) > 0

    if has_docker:
        result_sections.append(
            '## Docker (Optional)\n\nIf you prefer using Docker:\n```bash\ndocker-compose up -d\n```'
        )
    if has_database:
        dbs = ', '.join(tech_stack.get('databases', []))
        result_sections.append(
            f'## Database Setup\n\nThis project uses {dbs}. Ensure it is running locally or configure the connection in your environment variables.'
        )
    result_sections.append(
        '## Troubleshooting\n\nIf you encounter issues:\n1. Ensure all prerequisites are installed\n2. Clear your package manager cache and reinstall dependencies\n3. Check that environment variables are properly configured\n4. Look for existing GitHub Issues or open a new one'
    )
    return '\n\n'.join(result_sections)
